core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -545,14 +545,14 @@ final class ZSTM[-R, +E, +A] private[stm] (
   /**
    * Named alias for `<>`.
    */
-  def orElse[R1 <: R, E1, A1 >: A](that: => ZSTM[R1, E1, A1]): ZSTM[R1, E1, A1] =
+  def orElse[R1 <: R, E1 >: E, A1 >: A](that: => ZSTM[R1, E1, A1]): ZSTM[R1, E1, A1] =
```

Comment:
We have a problem with this: ZIO#orElse will try the right if the left fails. That's the "error swallowing" behavior you mention.

Maybe what this is saying is we need two combinators: one for doing "retry" fallback, and one for doing "error" fallback. What we used to have was one combinator for doing both. Or maybe doing two things at once is sufficiently common to warrant conflating them?

Reply:
This is something I briefly discussed with Adam. The current behavior of `orElse` (alternative) is wrong according to the cited paper and Haskell's implementation. The only thing I wouldn't do is leave it incorrect. Other than that, we are going to be asymmetric with something.